Add utilities to support handling of nested python data structures

zou3519 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #46289 Allow vmap to accept nested python data structures as inputs
* #46288 Implement `_broadcast_to_and_flatten(pytree, spec)`
* **#46287 Add utilities to support handling of nested python data structures**

This adds a lightweight `pytree` implementation that is similar to and
inspired by JAX pytrees, tensorflow.nest, deepmind/tree,
TorchBeast's TensorNest, etc.

A *pytree* is Python nested data structure. It is a tree in the sense
that nodes are Python collections (e.g., list, tuple, dict) and the leaves
are Python values. Furthermore, a pytree should not contain reference
cycles.

This PR:
- adds support for flattening and unflattening nested Python list/dict/tuples

Context: nested Tensor inputs for vmap
--------------------------------------
Right now, vmap is restricted to taking in flat lists of tensors. This
is because vmap needs to be able to convert every tensor in the input
that is being vmapped over into a BatchedTensor.

With a pytree library, we can simply flatten the input data structure
(returning the leaves), map all of the Tensors in the flat input to
BatchedTensors, and unflatten the flat list of BatchedTensors into a new
input. Or equivalently, with a `tree_map` function, we can map a nested
python data structure containing Tensors into one containing
BatchedTensors.

Future work
-----------
In some future PRs, we'll add nested input support for vmap. The
prerequisites for that are:
- a `broadcast_to(small, big)` that broadcasts `small` up to `big`.
  This is for handling the in_dims to vmap: the in_dims structure must
  be compatible with the structure of the inputs.

Test Plan
---------
- New tests in test/test_pytree.py

Differential Revision: [D24392890](https://our.internmc.facebook.com/intern/diff/D24392890)

zou3519 committed:
Add utilities to support handling of nested python data structures

This adds a lightweight `pytree` implementation that is similar to and
inspired by JAX pytrees, tensorflow.nest, deepmind/tree,
TorchBeast's TensorNest, etc.

A *pytree* is Python nested data structure. It is a tree in the sense
that nodes are Python collections (e.g., list, tuple, dict) and the leaves
are Python values. Furthermore, a pytree should not contain reference
cycles.

This PR:
- adds support for flattening and unflattening nested Python list/dict/tuples

Context: nested Tensor inputs for vmap
--------------------------------------
Right now, vmap is restricted to taking in flat lists of tensors. This
is because vmap needs to be able to convert every tensor in the input
that is being vmapped over into a BatchedTensor.

With a pytree library, we can simply flatten the input data structure
(returning the leaves), map all of the Tensors in the flat input to
BatchedTensors, and unflatten the flat list of BatchedTensors into a new
input. Or equivalently, with a `tree_map` function, we can map a nested
python data structure containing Tensors into one containing
BatchedTensors.

Future work
-----------
In some future PRs, we'll add nested input support for vmap. The
prerequisites for that are:
- a `broadcast_to(small, big)` that broadcasts `small` up to `big`.
  This is for handling the in_dims to vmap: the in_dims structure must
  be compatible with the structure of the inputs.

Test Plan
---------
- New tests in test/test_pytree.py

[ghstack-poisoned]

zou3519 committed:
Update on "Add utilities to support handling of nested python data structures"


This adds a lightweight `pytree` implementation that is similar to and
inspired by JAX pytrees, tensorflow.nest, deepmind/tree,
TorchBeast's TensorNest, etc.

A *pytree* is Python nested data structure. It is a tree in the sense
that nodes are Python collections (e.g., list, tuple, dict) and the leaves
are Python values. Furthermore, a pytree should not contain reference
cycles.

This PR:
- adds support for flattening and unflattening nested Python list/dict/tuples

Context: nested Tensor inputs for vmap
--------------------------------------
Right now, vmap is restricted to taking in flat lists of tensors. This
is because vmap needs to be able to convert every tensor in the input
that is being vmapped over into a BatchedTensor.

With a pytree library, we can simply flatten the input data structure
(returning the leaves), map all of the Tensors in the flat input to
BatchedTensors, and unflatten the flat list of BatchedTensors into a new
input. Or equivalently, with a `tree_map` function, we can map a nested
python data structure containing Tensors into one containing
BatchedTensors.

Future work
-----------
In some future PRs, we'll add nested input support for vmap. The
prerequisites for that are:
- a `broadcast_to(small, big)` that broadcasts `small` up to `big`.
  This is for handling the in_dims to vmap: the in_dims structure must
  be compatible with the structure of the inputs.

Test Plan
---------
- New tests in test/test_pytree.py

[ghstack-poisoned]

ezyang commented:
Can this be used to subsume the flatten/unflatten in torch/autograd/function.py for IOFunction?

ezyang commented on torch/utils/_pytree.py:
```diff
@@ -0,0 +1,150 @@
+from typing import NamedTuple, Callable, Any, Tuple, List, Dict
```
This looks pretty well typed. Consider adding it to the mypy-strict.ini list so we keep it well typed!

ezyang commented:
@zdevito I cc'ed you because ISTR you showing me that looked pretty similar

ezyang commented on torch/utils/_pytree.py:
```diff
@@ -0,0 +1,150 @@
+#   (returned by flatten_fn). It returns the collection by reconstructing
+#   it from the list and the context.
+Context = Any
+pytree_t = Any
```
`pytree_t` is not a very idiomatic name for a type in Python

ezyang commented on torch/utils/_pytree.py:
```diff
@@ -0,0 +1,150 @@
+NodeDef = NamedTuple('NodeDef', [
+    ('flatten_fn', flatten_fn_t),
+    ('unflatten_fn', unflatten_fn_t),
+])
```
protip, you can also write it as:

```
class NodeDef(NamedTuple):
  flatten_fn: flatten_fn_t
  unflatten_fn: unflatten_fn_t
```

zou3519 replied:
This makes it look a lot nicer, thanks for pointing that out!

ezyang commented on torch/utils/_pytree.py:
```diff
@@ -0,0 +1,150 @@
+    ('unflatten_fn', unflatten_fn_t),
+])
+
+SUPPORTED_NODES: Dict[Any, NodeDef] = {}
```
nit: `Type[Any]` here would be more descriptive instead of just `Any`.

ezyang commented on torch/utils/_pytree.py:
```diff
@@ -0,0 +1,150 @@
+    to reconstruct the pytree.
+    """
+    if _is_leaf(pytree):
+        return [pytree], LeafSpec()
```
This looks dead

ezyang reviewed (approved):
Cool. Maybe there's some other stuff in PyTorch we can replace with this.

albanD reviewed (commented):
Looks rely nice!

Are there any perf concerns with moving everything to use this?
@ezyang  we could add this to custom Function, Module bw hooks (when they're fixed https://github.com/pytorch/pytorch/pull/46163), backward/autograd.grad APIs, etc

ezyang commented:
> Are there any perf concerns with moving everything to use this?

Yeah, Richard did perf measurements on the top diff of this stack and this is noticeably slower (unsurprisingly). But I think as long as this doesn't slow down "core" codepaths it is not a big deal (and I definitely agree that moving it to C++ will speed it up).

ezyang commented on torch/utils/_pytree.py:
```diff
@@ -0,0 +1,150 @@
+        self.num_leaves: int = sum([spec.num_leaves for spec in children_specs])
+
+    def __repr__(self):
+        return f'TreeSpec({self.type.__name__}, {self.context}, {self.children_specs})'
```
Oh this is a good point; @zou3519 you should consider making this a dataclass, then you can print easily by converting the dataclass to a dictionary and then pretty-printing it.

zou3519 replied:
This is definitely not very readable for very nested TreeSpecs.

I'll handle this in a follow-up task (#46538), I'm not completely sure TreeSpec should be a dataclass (because there's some logic in `__init__` that tracks the number of leaves but that can just be moved outside of `__init__`)

ezyang replied:
You can use `__post_init__` for that, it's pretty easy

albanD commented:
> this is noticeably slower (unsurprisingly).

In what kind of use cases? In particular maybe we can have a fast path for the "basic" use case where it is already a flat collection of Tensor (the only acceptable thing right now).

zou3519 commented:
> > this is noticeably slower (unsurprisingly).
> 
> In what kind of use cases? In particular maybe we can have a fast path for the "basic" use case where it is already a flat collection of Tensor (the only acceptable thing right now).

This adds 10us for a function that takes in 4 arguments. There are ways to speed it up (add a fast path like you suggested, or move parts of the implementation to C++)

zou3519 committed:
Update on "Add utilities to support handling of nested python data structures"


This adds a lightweight `pytree` implementation that is similar to and
inspired by JAX pytrees, tensorflow.nest, deepmind/tree,
TorchBeast's TensorNest, etc.

A *pytree* is Python nested data structure. It is a tree in the sense
that nodes are Python collections (e.g., list, tuple, dict) and the leaves
are Python values. Furthermore, a pytree should not contain reference
cycles.

This PR:
- adds support for flattening and unflattening nested Python list/dict/tuples

Context: nested Tensor inputs for vmap
--------------------------------------
Right now, vmap is restricted to taking in flat lists of tensors. This
is because vmap needs to be able to convert every tensor in the input
that is being vmapped over into a BatchedTensor.

With a pytree library, we can simply flatten the input data structure
(returning the leaves), map all of the Tensors in the flat input to
BatchedTensors, and unflatten the flat list of BatchedTensors into a new
input. Or equivalently, with a `tree_map` function, we can map a nested
python data structure containing Tensors into one containing
BatchedTensors.

Future work
-----------
In some future PRs, we'll add nested input support for vmap. The
prerequisites for that are:
- a `broadcast_to(small, big)` that broadcasts `small` up to `big`.
  This is for handling the in_dims to vmap: the in_dims structure must
  be compatible with the structure of the inputs.

Test Plan
---------
- New tests in test/test_pytree.py

[ghstack-poisoned]

zou3519 committed:
Update on "Add utilities to support handling of nested python data structures"


This adds a lightweight `pytree` implementation that is similar to and
inspired by JAX pytrees, tensorflow.nest, deepmind/tree,
TorchBeast's TensorNest, etc.

A *pytree* is Python nested data structure. It is a tree in the sense
that nodes are Python collections (e.g., list, tuple, dict) and the leaves
are Python values. Furthermore, a pytree should not contain reference
cycles.

This PR:
- adds support for flattening and unflattening nested Python list/dict/tuples

Context: nested Tensor inputs for vmap
--------------------------------------
Right now, vmap is restricted to taking in flat lists of tensors. This
is because vmap needs to be able to convert every tensor in the input
that is being vmapped over into a BatchedTensor.

With a pytree library, we can simply flatten the input data structure
(returning the leaves), map all of the Tensors in the flat input to
BatchedTensors, and unflatten the flat list of BatchedTensors into a new
input. Or equivalently, with a `tree_map` function, we can map a nested
python data structure containing Tensors into one containing
BatchedTensors.

Future work
-----------
In some future PRs, we'll add nested input support for vmap. The
prerequisites for that are:
- a `broadcast_to(small, big)` that broadcasts `small` up to `big`.
  This is for handling the in_dims to vmap: the in_dims structure must
  be compatible with the structure of the inputs.

Test Plan
---------
- New tests in test/test_pytree.py

Differential Revision: [D24392890](https://our.internmc.facebook.com/intern/diff/D24392890)

[ghstack-poisoned]
